grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -0,0 +1,323 @@
+            handleError(Status.fromThrowable(t).augmentDescription(
+                "Failed to connect to LB server"));
+          }
+        },
+        executor);
+  }
+
+  @GuardedBy("lock")
+  private void startNegotiation() {
+    Preconditions.checkState(lbTransport != null, "currentLbTransport must be available");
+    logger.info("Starting LB negotiation");
+    LoadBalanceRequest initRequest = LoadBalanceRequest.newBuilder()
+        .setInitialRequest(InitialLoadBalanceRequest.newBuilder()
+            .setName(serviceName).build())
+        .build();
+    lbResponseObserver = new LbResponseObserver();
+    sendLbRequest(lbTransport, initRequest);
+  }
+
+  @VisibleForTesting  // to be mocked in tests
+  @GuardedBy("lock")
+  void sendLbRequest(ClientTransport transport, LoadBalanceRequest request) {
+    Channel channel = new SingleTransportChannel(transport, executor,
+        deadlineCancellationExecutor, serviceName);
+    LoadBalancerGrpc.LoadBalancerStub stub = LoadBalancerGrpc.newStub(channel);
+    lbRequestWriter = stub.balanceLoad(lbResponseObserver);
+    lbRequestWriter.onNext(request);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    handleError(error.augmentDescription("Name resolution failed"));
+  }
+
+  @Override
+  public void shutdown() {
+    synchronized (lock) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      if (lbRequestWriter != null) {
+        lbRequestWriter.onCompleted();
+      }
+      executor = SharedResourceHolder.release(GrpcUtil.SHARED_CHANNEL_EXECUTOR, executor);
+      deadlineCancellationExecutor = SharedResourceHolder.release(
+          GrpcUtil.TIMER_SERVICE, deadlineCancellationExecutor);
+    }
+  }
+
+  @Override
+  public void transportShutdown(
+      EquivalentAddressGroup addressGroup, ClientTransport transport, Status status) {
+    handleError(status.augmentDescription("Transport to LB server closed"));
+    synchronized (lock) {
+      if (transport == lbTransport) {
+        connectToLb();
+      }
+    }
+  }
+
+  private void handleError(Status error) {
+    FulfillmentBatch<ClientTransport> pendingPicksFulfillmentBatch;
+    StatusException statusException = error.asException();
+    synchronized (lock) {
+      lastError = statusException;
+      pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+    }
+    pendingPicksFulfillmentBatch.fail(statusException);
+  }
+
+  private class LbResponseObserver implements StreamObserver<LoadBalanceResponse> {
+    @Override public void onNext(LoadBalanceResponse response) {
+      logger.info("Got a LB response: " + response);
+      FulfillmentBatch<ClientTransport> pendingPicksFulfillmentBatch;
+      final RoundRobinServerList newRoundRobinServerList;
+      HashMap<SocketAddress, ResolvedServerInfo> newServerMap =
+          new HashMap<SocketAddress, ResolvedServerInfo>();
+      synchronized (lock) {
```

Comment:
Why does the lock need to be held for any of the work before `roundRobinServerList = newRoundRobinServerList`?